src/v/kafka/server/group.cc

```diff
@@ -1728,66 +1746,74 @@ group::begin_tx(cluster::begin_group_tx_request r) {
-    if (
-      fence_it != _fence_pid_epoch.end()
-      && r.pid.get_epoch() < fence_it->second) {
+    if (fence_it == _fence_pid_epoch.end()) {
+        // intentionally empty
+    } else if (r.pid.get_epoch() < fence_it->second) {
         vlog(
           _ctx_txlog.error,
           "pid {} fenced out by epoch {}",
           r.pid,
           fence_it->second);
         co_return make_begin_tx_reply(cluster::tx_errc::fenced);
+    } else if (r.pid.get_epoch() > fence_it->second) {
+        // TODO: proactively cancel an ongoing txn
     }
 
-    if (
-      fence_it == _fence_pid_epoch.end()
-      || r.pid.get_epoch() > fence_it->second) {
-        group_log_fencing_v0 fence{.group_id = id()};
-
-        // TODO: https://app.clubhouse.io/vectorized/story/2200
-        // include producer_id into key to make it unique-ish
-        // to prevent being GCed by the compaction
-        auto batch = make_tx_batch(
-          model::record_batch_type::tx_fence,
-          fence_control_record_v0_version,
-          r.pid,
-          std::move(fence));
-        auto reader = model::make_memory_record_batch_reader(std::move(batch));
-        auto e = co_await _partition->raft()->replicate(
-          _term,
-          std::move(reader),
-          raft::replicate_options(raft::consistency_level::quorum_ack));
-
-        if (!e) {
+    auto txseq_it = _tx_seqs.find(r.pid.get_id());
+    if (txseq_it != _tx_seqs.end()) {
+        if (r.tx_seq != txseq_it->second) {
+             vlog(
+              _ctx_txlog.warn,
+              "can't begin a tx {} with tx_seq {}: a producer id is already "
+              "involved in a tx with tx_seq {}",
+              r.pid,
+              r.tx_seq,
+              txseq_it->second);
+            co_return make_begin_tx_reply(cluster::tx_errc::unknown_server_error);
+        }
+        if (_prepared_txs.contains(r.pid)) {
             vlog(
-              _ctx_txlog.error,
-              "Error \"{}\" on replicating pid:{} fencing batch",
-              e.error(),
-              r.pid);
+              _ctx_txlog.warn,
+              "can't begin a tx {} with tx_seq {}: it was already begun and it "
+              "accepted writes",
+              r.pid,
+              r.tx_seq);
             co_return make_begin_tx_reply(
               cluster::tx_errc::unknown_server_error);
         }
+        co_return cluster::begin_group_tx_reply(_term, cluster::tx_errc::none);
+    }
 
-        // _fence_pid_epoch may change while the method waits for the
-        //  replicate coroutine to finish so the fence_it may become
-        //  invalidated and we need to grab it again
-        fence_it = _fence_pid_epoch.find(r.pid.get_id());
-        if (fence_it == _fence_pid_epoch.end()) {
-            _fence_pid_epoch.emplace(r.pid.get_id(), r.pid.get_epoch());
-        } else {
-            fence_it->second = r.pid.get_epoch();
-        }
+    std::optional<model::record_batch> batch{};
+    if (is_transaction_ga()) {
```

Comment:
nit: no implications of is_transaction_ga() changing between scheduling points? (the issue Noah flagged), just want to double check.

Reply:
good point, I cache `is_transaction_ga` now